### ISSUE TITLE
Attribute to return type instead of Match pattern

### ISSUE THREAD

**Issue by SergeyKanzhelev** [open-telemetry/opentelemetry-dotnet#12]
https://github.com/open-telemetry/opentelemetry-java/pull/31

**Comment by trwegner** [comment]
Working on this

**Comment by austinlparker** [comment]
Spent a bit looking at this and had some questions - Right now this is an example of how `Match` gets used:

```
kvp => kvp.Value.Match(
                    s => new OpenTelemetry.Proto.Trace.V1.AttributeValue { StringValue = new TruncatableString() { Value = s } },
                    b => new OpenTelemetry.Proto.Trace.V1.AttributeValue { BoolValue = b },
                    l => new OpenTelemetry.Proto.Trace.V1.AttributeValue { IntValue = l },
                    d => new OpenTelemetry.Proto.Trace.V1.AttributeValue { DoubleValue = d },
                    o => new OpenTelemetry.Proto.Trace.V1.AttributeValue { StringValue = new TruncatableString() { Value = o?.ToString() } }
```
The Java implementation replaced this with a `getType` method and an enum, but they also had separate `AttributeValue` classes such as `AttributeValueLong`, etc. which don't exist in our case as `AttributeValue{T}` replaces them. I'm generally in favor of using generics where they make sense, and... well, it makes sense the way we're doing it right now, and `IAttributeValue` is already decoupled. Thoughts or suggestions?

**Comment by SergeyKanzhelev** [comment]
We need to switch to `getType`. Java used to have match pattern in OpenCensus as well and we all agreed that `match`'s benefits are not worth the complexity

**Comment by austinlparker** [comment]
Do we want to factor out `AttributeValue{T}`, then, like it is in Java?

**Comment by SergeyKanzhelev** [comment]
Yes. It may have been a mistake =)

**Comment by carlosalberto** [comment]
I don't think you guys *have* to switch to the `getType()` style. We did it in Java as this pattern does not work fine (lambdas exist only from late versions, while in C# they have existed for long time).

I'd also could imagine something in between, btw - something using generics (which are properly implemented in .Net) while exposing information about the type.

**Comment by bogdandrutu** [comment]
Generics in Java are a hustle, forcing boxing for primitive types. If that is not the case and we can have a nice pattern for add and retrieve then we don't need this.

**Comment by SergeyKanzhelev** [comment]
If keeping match - it would be great to fix the pattern. https://github.com/census-instrumentation/opencensus-csharp/issues/97

**Comment by SergeyKanzhelev** [comment]
Copying from liked issue:

This repository uses pattern matching with some errors:

The purpose of `Match` method is to avoid `if/else` checks by supplying set of callbacks to call. And each implementation calls respective callback. Something like a virtual methods in generic class, but strongly typed for pre-defined types.

Issues:

1. `AttributeValue` uses if/else instead of having separate classes with the hardcoded implementation of a `Match` method which will call the proper callback: 
https://github.com/census-instrumentation/opencensus-csharp/blob/d087855fb93440f678009173d3f0f4392b58ce52/src/OpenCensus.Abstractions/Trace/Internal/AttributeValue%7BT%7D.cs#L120-L149

2. There is no need to expose type-specific results in factory methods `AttributeValue.StringAttributeValue`. Simply returning `IAttributeValue` should suffice. So generic `IAttributeValue<T>` interface is not needed.

3. Implementation of IAttributeValue is trivial and there is no need to have interface and implementation separately - simple implementation class will suffice.

**Comment by austinlparker** [comment]
Do you have anything I can look at that's closer to the desired implementation? I'm having trouble understanding what we're looking to do here.

**Comment by SergeyKanzhelev** [comment]
1. `Match` is often used to avoid the long `if/else` by using polimorpism. Like here https://github.com/open-telemetry/opentelemetry-dotnet/blob/bca11b84189de0d92c6ea2206a16ff0d0312427c/src/OpenTelemetry/Stats/Aggregations/Mean.cs#L34-L37

2. There is no need to return something different than a generic `Attribute` from the calls like `AttributeValue.StringAttributeValue`. The way `AttributeValue.StringAttributeValue` we need to expose the details on what specific type will be exposed by this call, thus adding more public classes and interfacves to the API surface.

3. There is no situation AttributeValue needs to be an interface. In Java reference API AttributeValue is an abstract class. I might suggest to go even further and simply make is a final class (this is optional).

**Comment by austinlparker** [comment]
Closed by #88 #148 #149
